#162948843 Implement Rating of Articles

andrewhingah opened the pull request:
**What does this PR do?**

Implements rating of articles where ratings are integers from 1 to 5.

**Description of tasks to be completed**

- create article rating model
- create an article rating serializer class
- add view for viewing an article's average ratings
- add view for posting a rating to an article
- refactor tests

**How should this be manually tested?**

- `git clone https://github.com/andela/ah-django-unchained.git`
- `git checkout ft-article-rating-162948843`
- `source .env`
- `python migrate`
- run tests `python manage.py test`

On Postman, register 2 or more users via the endpoint `http://127.0.0.1:8000/api/users/`.

Create an article while signed in via the endpoint `http://127.0.0.1:8000/api/articles/` using one of the users details.

To rate the article:
Make a POST request on `http://127.0.0.1:8000/api/articles/<slug>/rate/`. Ensure to copy the token of one of the other user and add it as an authorization header in the request.
The request body should look like this;
`
{
	"rate": 2
}
`
The response is a message that the article has been successfully rated.

To view the ratings of a specific article:
Make a GET request on `http://127.0.0.1:8000/api/articles/rate/<slug>/`.
The average ratings of that article should be displayed in the response.

**Any background context you want to provide?**

A user cannot rate their own article

**Relevant Pivotal Tracker Story**

[#162948843](https://www.pivotaltracker.com/story/show/162948843)

**Screenshots**

Rate article
<img width="976" alt="screenshot 2019-01-28 at 09 50 16" src="https://user-images.githubusercontent.com/32644295/51819187-91779f80-22e2-11e9-85f8-c066aceaa462.png">




View article ratings
<img width="981" alt="screenshot 2019-01-25 at 09 17 51" src="https://user-images.githubusercontent.com/32644295/51728691-2a5bb000-2082-11e9-8be3-5dfb3910ea15.png">

Give a rating of more than 5
<img width="963" alt="screenshot 2019-01-27 at 12 23 33" src="https://user-images.githubusercontent.com/32644295/51799119-7d726600-222e-11e9-9ae0-e09d32a4c443.png">

andrewhingah committed:
feat(article rating): add unit tests for article rating

- write unit tests for article rating

[Delivers #162948843]

andrewhingah committed:
feat(article rating): implement rating of articles

- create article rating model
- create an article rating serializer class
- add view for viewing an article's average ratings
- add view for posting a rating to an article
- refactor tests

[Delivers #162948843]

Georgeygigz commented on authors/apps/articles/tests/test_article_ratings.py:
```diff
@@ -0,0 +1,97 @@
+        slug = self.slug
+        data = {"rate": 4}
+        response = self.client.post(
+            "/api/articles/{}/rate/".format(slug), data=data, format='json')
```
assert for the content of the response too

Georgeygigz commented on authors/apps/articles/tests/test_article_ratings.py:
```diff
@@ -0,0 +1,97 @@
+        data = {"rate": 3}
+        response = self.client.post(
+            "/api/articles/{}/rate/".format(slug), data=data)
+        self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
```
ditto

Georgeygigz commented on authors/apps/articles/tests/test_article_ratings.py:
```diff
@@ -0,0 +1,97 @@
+        data = {"rate": 4}
+        response = self.client.post(
+            "/api/articles/{}/rate/".format(slug), data=data, format='json')
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
```
ditto

Georgeygigz commented on authors/apps/articles/tests/test_article_ratings.py:
```diff
@@ -0,0 +1,97 @@
+class TestArticleRatings(APITestCase):
+    def setUp(self):
+        self.client = APIClient()
+        self.register_url = reverse('authentication:auth-register')
```
This is not used

Georgeygigz commented on authors/apps/articles/views.py:
```diff
@@ -56,3 +60,85 @@ class DeleteArticle(UpdateAPIView):
+
+        user = request.user
+        article_ratings = None
+        res = Response(
```
kindly replace (res) with a descriptive variable name

SelaDanti commented on authors/apps/articles/urls.py:
```diff
@@ -1,14 +1,22 @@
 from django.urls import path
 
```
Add an extra line after imports

SelaDanti commented on authors/apps/articles/urls.py:
```diff
@@ -1,14 +1,22 @@
+from .views import (
+     ArticleAPIView, ArticleDetailsView,
+     DeleteArticle, PostRatingsAPIView, AverageRatingsAPIView
+     )
```
ditto

ignatiusukwuoma commented on authors/apps/articles/serializers.py:
```diff
@@ -1,7 +1,7 @@
 
 from rest_framework import serializers
-from .models import Article
+from .models import Article, ArticleRating
```
ArticleRating is not being used in this file

andrewhingah replied:
Let me work on that.

ignatiusukwuoma commented on authors/apps/articles/urls.py:
```diff
@@ -14,5 +18,11 @@
+         name="rate_article"),
+    path('articles/rate/<slug>/',
+         AverageRatingsAPIView.as_view(),
+         name="articles-view_ratings")
```
Maybe you should use only underscores or hyphens

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -2,20 +2,24 @@
 
 from django.template.defaultfilters import slugify
 from django.core.exceptions import ObjectDoesNotExist
+from rest_framework import status
```
duplicate import

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -131,3 +135,85 @@ def put(self, request, slug):
+        try:
+            rate_articles = ArticleRating.objects.filter(article_id=art_id)
+        except Exception as e:
+            raise serializers.ValidationError(str(e))
```
Please delete the import of `serializers` from line 18

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -131,3 +135,85 @@ def put(self, request, slug):
+        art_id = article.id
+
+        try:
+            rate_articles = ArticleRating.objects.filter(article_id=art_id)
```
Just filter using `article` instead of `art_id`. i.e `article=article`.

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -131,3 +135,85 @@ def put(self, request, slug):
+
+    def get(self, request, slug):
+        """method for viewing average ratings"""
+        rate_article = None
```
Not necessary... and the variable is not even been used, `rate_articles` is

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -131,3 +135,85 @@ def put(self, request, slug):
+            rate += rate_article.rate
+        rate_value = 0
+        if rate:
+            rate_value = rate / len(rate_articles)
```
I thought we have discussed a more efficient way of finding the length of a queryset

andrewhingah replied:
Noted. Will not repeat that.

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -131,3 +135,85 @@ def put(self, request, slug):
+        """method for posting a rating"""
+        data = request.data
+        serializer = self.serializer_class(data=data)
+        serializer.validate(data=data)
```
You don't need this to validate the serializer. What is it doing?
Context: https://www.django-rest-framework.org/api-guide/serializers/#validation

andrewhingah replied:
Let me get rid of it

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -131,3 +135,85 @@ def put(self, request, slug):
+        try:
+            article = Article.objects.get(slug=slug)
+        except ObjectDoesNotExist:
+            raise NotFound('article not found')
```
How about starting with a capital letter

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -131,3 +135,85 @@ def put(self, request, slug):
+        author = article.author.id
+        if author == request.user.id:
+            return Response(
+                {"response": "You are not allowed to rate your own article"},
```
Shouldn't the object key be `error` instead of 'response'?

andrewhingah replied:
Noted

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -131,3 +135,85 @@ def put(self, request, slug):
+        user = request.user
+        article_ratings = None
+        response = Response(
+                {"response": "article successfully rated"},
```
ditto... start with capital letter. return the rating as response and `message` as the object key

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -131,3 +135,85 @@ def put(self, request, slug):
+            article_ratings.save()
+            return response
+        # create new rating if none exists
+        except Exception:
```
ditto

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -131,3 +135,85 @@ def put(self, request, slug):
+
+        try:
+            rate_articles = ArticleRating.objects.filter(article_id=art_id)
+        except Exception as e:
```
Exception clause is too broad.

ignatiusukwuoma commented on authors/apps/articles/tests/test_article_ratings.py:
```diff
@@ -0,0 +1,154 @@
+from django.urls import reverse
+from rest_framework.test import APIClient, APITestCase
+from rest_framework import status
+from authors.apps.articles.models import Article
```
this import is not being used

andrewhingah committed:
feat(article rating): remove unused imports, use count, refactor error messages

- remove unused imports
- use count instead of len to get the length of queryset
- refactor error messages
- get rid of unnecessary exception

[Fixes #162948843]

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -131,3 +133,86 @@ def put(self, request, slug):
+            article_ratings.save()
+            return response
+        # create new rating if none exists
+        except Exception:
```
How about a `DoesNotExist` exception?

andrewhingah replied:
How about I use `ObjectDoesNotExist` on this one?

andrewhingah replied:
I have realised `DoesNotExist` works well. Thanks.

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -131,3 +133,86 @@ def put(self, request, slug):
+        response = Response(
+                {
+                    "message": "Article successfully rated",
+                    "rating": serializer.data
```
I think the data is over-nested. How about `rate: serializer.data.rate`

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -131,3 +133,86 @@ def put(self, request, slug):
+                    "message": "Article successfully rated",
+                    "rating": serializer.data
+                },
+                status=status.HTTP_200_OK
```
Shouldn't it be a 201?

andrewhingah committed:
feat(article rating): use ObjectDoesNotExist exception, return a non-nested rating

- use ObjectDoesNotExist exception where article rating is unavailable
- return an unnested rating value in response
- return status 201 in create requests instead of 200

[Fixes #162948843]

andrewhingah committed:
feat(article rating): use DoesNotExist where no ratings exist

- use DoesNotExist exception where no ratings exist

[Fixes #162948843]
